### ISSUE TITLE
[bug] add attachment function not defined

### ISSUE THREAD

**Issue by aMOPel** [pimalaya/himalaya-vim#9]
Thank you for the great work on himalaya!

## Problem

`ga` gives error `E117: Unknown function: himalaya#domain#email#add_attachment`

and I could not find any references in the repo to any function relating to adding an attachment.

Did it get lost somewhen?

**Comment by soywod** [comment]
The doc seems not up-to-date with the latest version. I also need to adjust the plugin with the `v1.0.0` of the CLI (actually being tested on `master`). Thank you for pointing it out!

Regarding attachments, I would advise you to read the [FAQ](https://github.com/pimalaya/himalaya?tab=readme-ov-file#faq): *How to add attachments to a message?*. Attachments are added directly from the message composition.

**Comment by aMOPel** [comment]
It's not just the docs, though. You are also creating the `ga` mapping. It just points to a non existent function.

> Regarding attachments, I would advise you to read the FAQ: How to add attachments to a message?. Attachments are added directly from the message composition.

I saw that part. I just thought you meant to provide a mapping to conveniently put that boilerplate text, since it is a common operation.

Could be done by directly writing the boilerplate into the mapping or with a snippet (see https://github.com/rafamadriz/friendly-snippets for cross plugin snippets).

Also the file picking for the attachment could be integrated with telescope for example. 

Just some ideas for little niceties.

Any way thank you again for your work!

**Comment by soywod** [comment]
> It's not just the docs, though. You are also creating the `ga` mapping. It just points to a non existent function.

Yes indeed, the code is also not up-to-date. The whole repo needs to be adjusted.

> > Regarding attachments, I would advise you to read the FAQ: How to add attachments to a message?. Attachments are added directly from the message composition.
> 
> I saw that part. I just thought you meant to provide a mapping to conveniently put that boilerplate text, since it is a common operation.

Attachments used to be attached separately from the message, which is not the case anymore. I thought to re-use the function to interactively put the right code inside the body (like asking the path to the file, the MIME type, the description etc). Could be a great addition.

> Could be done by directly writing the boilerplate into the mapping or with a snippet (see https://github.com/rafamadriz/friendly-snippets for cross plugin snippets).

Also, but this should remain user side. I prefer not to deal with them, since there are too many engines out there.

> Also the file picking for the attachment could be integrated with telescope for example.

Exactly!

**Comment by aMOPel** [comment]
> Also, but this should remain user side. I prefer not to deal with them, since there are too many engines out there.

There are many engines, but there are only 2 bigger standards, one of which is at least supported by a snippet engine, either snipmate-like or vscode-like.

Snipmate-like snippets are collected here
https://github.com/honza/vim-snippets

And vscode-like snippets are collected here https://github.com/rafamadriz/friendly-snippets

Making one of each would suffice to be compatible with, I believe all, vim snippet engines. And those standards are stable afaik know, so it's a one time effort.

**Comment by aMOPel** [comment]
actually i made it now:

```lua
local pick_and_paste_filepath_at = function(root_dir)
  if root_dir == nil then
    root_dir = "~"
  end

  if type(root_dir) ~= "string" then
    vim.notify(
      "pick_and_paste_filepath_at root_dir needs to be string",
      vim.log.levels.ERROR
    )
    return
  end

  root_dir = vim.fs.normalize(root_dir)

  -- disable BufLeave autocmd temporarily
  local au_opts = {
    group = "himalaya_write",
    event = { "BufLeave" },
  }
  local au = vim.api.nvim_get_autocmds(au_opts)
  if au == nil or #au ~= 1 then
    vim.notify(
      "pick_and_paste_filepath_at: expected exactly one autocmd but got: \n"
        .. vim.inspect(au),
      vim.log.levels.ERROR
    )
    return
  end
  au = au[1]
  vim.api.nvim_clear_autocmds(au_opts)

  local actions = require("telescope.actions")
  local action_state = require("telescope.actions.state")

  -- save working directory to restore later
  local prev_wd = vim.fn.getcwd()

  -- cd to root_dir only for window
  vim.cmd(string.format("silent lcd %s", vim.fn.expand(root_dir)))

  local function run_selection(prompt_bufnr, map)
    -- replace default action on <cr> in telescope picker
    actions.select_default:replace(function()
      -- close telescope buffer
      actions.close(prompt_bufnr)
      -- insert selected filepath
      local selection = action_state.get_selected_entry()
      vim.paste({ vim.fs.joinpath(root_dir, selection[1]) }, -1)
      -- restore wd for window
      vim.cmd(
        string.format("silent lcd %s", vim.fn.expand(prev_wd))
      )
      -- restore BufLeave autocmd
      vim.api.nvim_create_autocmd(au.event, {
        group = au.group,
        buffer = au.buffer,
        command = au.command,
        once = au.once,
      })
      -- restore cursor position
      vim.cmd([[normal! 'z]])
    end)
    -- has to return true
    return true
  end

  -- run telescope file picker with our custom function
  local opts = { attach_mappings = run_selection }
  require("telescope.builtin").find_files(opts)
end

-- add attachment at the end of file, pick file with telescope at `root_dir`
local add_attachment = function(root_dir)
  local k = vim.keycode
  return function()
    -- set mark to restore cursor postion later
    vim.cmd([[normal! mz]])
    -- set goto end and add 2 newlines
    vim.cmd([[normal! Go]] .. k("<cr>") .. k("<esc>"))
    -- insert attachment boilerplate
    vim.paste({ "<#part filename=><#/part>" }, -1)
    -- move cursor to =
    vim.cmd([[normal! 0f=]])
    pick_and_paste_filepath_at(root_dir)
  end
end

vim.api.nvim_create_autocmd({ "FileType" }, {
  group = "MyAutoCmd",
  pattern = { "mail" },
  callback = function(ev)
    vim.keymap.set("n", "gtt", add_attachment("~"), {
      desc = "add email attachment",
      buffer = true,
      silent = true,
    })
  end,
  desc = "add email attachment",
})
```

adapt as you need.

it puts the file attachment at the end of the buffer.

it got the telescope file picker, you can pass the root dir for the file search, I figured `~` would be good.

it restores the cursor position.

it temporarily removes the autocmd on bufleave, when swiching to the telescope buffer.

**Comment by soywod** [comment]
Unfortunately no update, I still work on low-level libs. I still need months. Feel free to propose a PR!

Alternatively, you may want to get closer to this person: https://github.com/pimalaya/himalaya/issues/580#issuecomment-3064431626

**Comment by mxaddict** [comment]
Thanks, I'll check it out.

Might make sense to port the current simple client version that you have
to also use lua, the version that you linked seems to be more complex
and is more of a traditional client instead of a light front-end like
the current VIM plugin that you have.

I'm guesing the reason for the complexity that he has in his version is
to have a more responsive feeling client as in it can load messages in
the background and store locally for faster interactions on the tui.
